### PR TITLE
directory: improve support for empty QueryResult (bug 1790047)

### PR DESCRIPTION
**Pull request by zzzeid**
- add `QueryResult.__bool__`
- make `result` and `rejected` parameters optional
- fix bug when adding rejected paths

### CODE REVIEW

**Review by dklawren — APPROVED** [review]
LGTM and testing worked as expect. r=dkl

**Review by cgsheeh — CHANGES_REQUESTED** [review]

File: tests/test_directory.py [hunk 1]
```diff
@@ -139,3 +140,42 @@ def test_directory__Directory__query_merging(repo):
...
+
+def test_directory__QueryResult_empty():
+    empty_result = QueryResult()
+    assert not empty_result
+
+
+def test_directory__QueryResult_empty_addition():
+    empty_result = QueryResult()
+    other_empty_result = QueryResult()
+    assert not (empty_result + other_empty_result)
```
[the remark this answers is not in the document]
Comment (zzzeid) [reply tests/test_directory.py:181]: That codepath is implicitly tested as part of `test_directory__Directory__query` and `test_directory__Directory__query_merging`, though I could add an explicit test here to test the basic case of non-empty `QueryResult`.

**Review by cgsheeh — APPROVED** [review]
